### PR TITLE
Function setoid is back.

### LINKED ISSUE #1467: Why does equality of Function Setoids have cong baked in?
Right now, we define function setoids as follows (taken from [Function.Equality](https://github.com/agda/agda-stdlib/blob/master/src/Function/Equality.agda#L79)):
```agda
setoid : ∀ {f₁ f₂ t₁ t₂}
         (From : Setoid f₁ f₂) →
         IndexedSetoid (Setoid.Carrier From) t₁ t₂ →
         Setoid _ _
setoid From To = record
  { Carrier       = Π From To
  ; _≈_           = λ f g → ∀ {x y} → x ≈₁ y → f ⟨$⟩ x ≈₂ g ⟨$⟩ y
  ; isEquivalence = record
    { refl  = λ {f} → cong f
    ; sym   = λ f∼g x∼y → To.sym (f∼g (From.sym x∼y))
    ; trans = λ f∼g g∼h x∼y → To.trans (f∼g From.refl) (g∼h x∼y)
    }
  }
  where
  open module From = Setoid From using () renaming (_≈_ to _≈₁_)
  open module To = IndexedSetoid To   using () renaming (_≈_ to _≈₂_)
```

If we look at the definition of `_≈_`, we can see that it has `cong` baked into it. This makes working with these equalities awkward, as you essentially perform a proof that `f ⟨$⟩ x ≈₂ g ⟨$⟩ x`, then perform a `cong` at the very end to get that `g ⟨$⟩ x ≈₂ g ⟨$⟩ y`. It seems like the following equivalent definition would be a bit easier to work with:
```agda
setoid : ∀ {f₁ f₂ t₁ t₂}
         (From : Setoid f₁ f₂) →
         IndexedSetoid (Setoid.Carrier From) t₁ t₂ →
         Setoid _ _
setoid From To = record
  { Carrier       = Π From To
  ; _≈_           = λ f g → ∀ x → f ⟨$⟩ x ≈₂ g ⟨$⟩ x
  ; isEquivalence = record
    { refl  = λ {f} _ → To.refl
    ; sym   = λ f∼g x → To.sym (f∼g x)
    ; trans = λ f∼g g∼h x → To.trans (f∼g x) (g∼h x)
    }
  }
  where
  open module From = Setoid From using () renaming (_≈_ to _≈₁_)
  open module To = IndexedSetoid To   using () renaming (_≈_ to _≈₂_)
```
 
 Thoughts?

### LINKED ISSUE #2228: Equality of functions in the new function hierarchy
As a counterpart to the now-deprecated `Function.Equality.setoid`.

Briefly discussed previously in https://github.com/agda/agda-stdlib/issues/1467#issuecomment-816737487

### PR DESCRIPTION
**Pull request by JacquesCarette**
Fixes #2228 . Fixes #1467 .

### CODE REVIEW

File: src/Function/Relation/Binary/Equality.agda
Comment (MatthewDaggitt) [inline src/Function/Relation/Binary/Equality.agda:9]: So it would be great if we could structure this better than the old hierarchy. In particular this should be `Equality.Setoid` so we have room for the propositional version as well.

File: src/Function/Relation/Binary/Equality.agda
Comment (MatthewDaggitt) [inline src/Function/Relation/Binary/Equality.agda:19]: Can we parameterise the whole module by the two setoids?

File: src/Function/Relation/Binary/Equality.agda
Comment (MatthewDaggitt) [inline src/Function/Relation/Binary/Equality.agda:22]: Given that we're parameterising the whole module, we should then be able to define a nice infix name for the raw equality relation.

File: src/Function/Relation/Binary/Equality.agda
Comment (MatthewDaggitt) [inline src/Function/Relation/Binary/Equality.agda:24]: And the proofs of reflexivity, symmetry, transitivity can be lifted out as their own lemmas.

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]

File: src/Function/Relation/Binary/Equality.agda
Comment (JacquesCarette) [reply src/Function/Relation/Binary/Equality.agda:9]: LOL, I named this according to your own suggestion in one of the issues! But sure, will do.

File: src/Function/Relation/Binary/Equality.agda
Comment (JacquesCarette) [reply src/Function/Relation/Binary/Equality.agda:24]: All done - though it's not quite as nice since I have to eta-expand.

File: src/Function/Relation/Binary/Setoid/Equality.agda
Comment (MatthewDaggitt) [inline src/Function/Relation/Binary/Setoid/Equality.agda:15]: We don't usually indent after top level modules?

File: src/Function/Relation/Binary/Setoid/Equality.agda
Comment (MatthewDaggitt) [inline src/Function/Relation/Binary/Setoid/Equality.agda:25]: We could get away with removing thistype annotation if we used a forall?

**Review by MatthewDaggitt — COMMENTED** [review]
Other than that looks good! Apologies for being inconsistent with my past self!

File: src/Function/Relation/Binary/Setoid/Equality.agda
Comment (JacquesCarette) [reply src/Function/Relation/Binary/Setoid/Equality.agda:15]: Indeed - that's one small difference between stdlib and categories. Can fix.

File: src/Function/Relation/Binary/Setoid/Equality.agda
Comment (JacquesCarette) [reply src/Function/Relation/Binary/Setoid/Equality.agda:25]: We could, but I rather think that this one serves as useful documentation.

Comment (MatthewDaggitt) [reply src/Function/Relation/Binary/Setoid/Equality.agda:25]: Okay sure.

Comment (jamesmckinna) [reply src/Function/Relation/Binary/Setoid/Equality.agda:25]: Any reason to favour the 'diagonal' version of the relation, rather than the PER-like
```agda
f ≈ g = {x y : From.Carrier} → x From.≈ y → f ⟨$⟩ x To.≈ g ⟨$⟩ y
```
(which, incidentally/moreover, would also permit use of a forall...)

Comment (JacquesCarette) [reply src/Function/Relation/Binary/Setoid/Equality.agda:25]: I forget the exact places where the discussion was had, but this is an explicit choice. The seeming extra generality doesn't buy you anything as these two are inter-derivable and it seems that in practice, it is this version which is more often needed.

Certainly we went back-and-forth on this in `agda-categories` and the version here definitely makes things simpler, at no loss of generality.

**Review by MatthewDaggitt — COMMENTED** [review]
Okay once the indentation is fixed, happy to merge this in.

**Comment by JacquesCarette** [comment]
Ok @MatthewDaggitt this should now be ready to go.

File: src/Function/Relation/Binary/Setoid/Equality.agda
Comment (jamesmckinna) [reply src/Function/Relation/Binary/Setoid/Equality.agda]: Yes, I now see the discussion on #1467 and while I may *personally* disagree, ... whatever!?

**Review by Taneb — APPROVED** [review]

**Comment by jamesmckinna** [comment]
I'd be grateful for the equivalence of the `cong`/PER-style definition and the proposed one, but happy to follow this PR up with another one, perhaps as part of tackling #2299 so I've no wish to hold this one up. But the `CHANGELOG` merge conflict is managing that on its own... happy to fix that if you don't have time @JacquesCarette ...

**Comment by JacquesCarette** [comment]
`CHANGELOG` conflict fixed and all checks have passed.

**Comment by JacquesCarette** [comment]
@MatthewDaggitt if you could please review? Your 'request changes' is holding this PR back from being merged.

**Comment by jamesmckinna** [comment]
Suggest dismissing the earlier review and approving the revised version instead...
@MatthewDaggitt would you be happy with that?

**Comment by jamesmckinna** [comment]
I don't think this PR, given its size and uncontroversial nature (I'm happy with the design rationale), needs either a second reviewer, nor to wait on unlocking the old stale review changes, so adding to the merge queue now in the interests of unblocking the `CHANGELOG` hell it has already been subject to.

**Pull request merged**
